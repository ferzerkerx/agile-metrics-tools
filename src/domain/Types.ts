import { MetricItem } from './metrics';

export interface TeamMetricsRequest {
  since?: Date;
  until?: Date;
  shouldUpdateEntries: boolean;
  config: any;
}

export interface ElasticSearchService {
  pushMetrics<T extends MetricItem>(
    metricItems: T[],
    shouldReplaceEntry?: boolean
  ): Promise<void>;
}
