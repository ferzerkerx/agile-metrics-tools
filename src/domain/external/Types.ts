export interface ExternalConfig {
  since: Date;
  until?: Date;
  type: string;
  srcType: string;
  inlineData: ExternalData[];
}

export class ExternalData {
  createdAt: Date;
  teamName: string;
  [type: string]: any;
}

export interface ExternalService {
  fetch(externalConfig: ExternalConfig): Promise<ExternalData[]>;
}

export interface ExternalRepository {
  csv(externalConfig: ExternalConfig): Promise<ExternalData[]>;
}
