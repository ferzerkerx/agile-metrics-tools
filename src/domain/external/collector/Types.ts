export class ExternalCollectorConfig {
  since: Date;
  until?: Date;
  type: string;
  srcType: string;
  metricType: string;
  inlineData: [];
  constructor({
    since,
    until = null,
    type,
    srcType,
    metricType,
    inlineData = null,
  }) {
    this.since = since ? new Date(since) : null;
    this.until = until ? new Date(until) : null;
    this.type = type;
    this.srcType = srcType;
    this.metricType = metricType;
    this.inlineData = inlineData;
  }
}
