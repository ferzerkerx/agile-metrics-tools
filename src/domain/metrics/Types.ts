import { ExternalCollectorConfig } from '../external/collector/Types';

export interface CollectorService {
  fetch(config: ExternalCollectorConfig): Promise<MetricItem[]>;
}

export interface MetricItem {
  id: string;
  dataType: string;
  createdAt: Date;
  teamName: string;
}
