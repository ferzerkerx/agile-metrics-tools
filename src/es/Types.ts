import { MetricItem } from '../metrics';

export interface ElasticSearchService {
  push(payload: MetricItem): Promise<any>;
}
