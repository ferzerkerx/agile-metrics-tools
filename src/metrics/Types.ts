export interface CollectorService<
  T extends CollectorConfig,
  K extends MetricItem
> {
  fetch(config: T): Promise<K[]>;
  supports(config: any): boolean;
}

export interface MetricItem {
  id: string;
  dataType: string;
  createdAt: Date;
  teamName: string;
}

export interface CollectorConfig {
  teamName: string;
}

export interface CollectorModuleFactory<
  T extends CollectorConfig,
  K extends MetricItem
> {
  collectorInstance(): CollectorService<T, K>;
  collectorConfiguration(obj: any): T;
}
