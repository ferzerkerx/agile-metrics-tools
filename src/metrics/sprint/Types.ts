export interface Statistics {
  moveBackward: number;
  moveForward: number;
  movedToDev: Date;
}
