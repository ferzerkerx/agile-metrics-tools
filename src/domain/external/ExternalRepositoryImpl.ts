import { ExternalConfig, ExternalData, ExternalRepository } from './Types';
import { Converters } from './Converters';
import { Utils } from '../metrics';

export class ExternalRepositoryImpl implements ExternalRepository {
  async csv(externalConfig: ExternalConfig): Promise<ExternalData[]> {
    if (!ExternalRepositoryImpl.isDate(externalConfig.since)) {
      externalConfig.since = new Date(externalConfig.since);
    }
    return Promise.resolve(
      this.relevantExternalData(externalConfig, externalConfig.inlineData)
    );
  }

  private static isDate(date: any): boolean {
    return date instanceof Date && !isNaN(date.valueOf());
  }

  private relevantExternalData(
    externalConfig: ExternalConfig,
    externalConfigurations: ExternalData[]
  ): ExternalData[] {
    return externalConfigurations
      .map((jsonObj) => Converters.toExternalData(jsonObj))
      .filter((item) =>
        Utils.isDateInRange({
          createdAt: item.createdAt,
          until: externalConfig.until,
          since: externalConfig.since,
        })
      );
  }
}
