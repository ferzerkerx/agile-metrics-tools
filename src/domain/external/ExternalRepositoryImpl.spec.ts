import { ExternalRepositoryImpl } from './ExternalRepositoryImpl';
import { ExternalConfig, ExternalRepository } from './Types';

describe('ExternalRepositoryImpl', () => {
  const externalRepository: ExternalRepository = new ExternalRepositoryImpl();

  it('should get inline values properly', async () => {
    const externalConfig: ExternalConfig = {
      since: new Date('2018-12-03'),
      type: 'csv',
      srcType: 'inline',
      inlineData: [
        {
          createdAt: new Date('2018-12-04T00:00:00.000Z'),
          teamName: 'someTeamName',
          header1: 2,
          header2: 3,
          header3: 'someValue',
        },
      ],
    };

    const externalData = await externalRepository.csv(externalConfig);
    expect(externalData).toMatchSnapshot();
  });
});
