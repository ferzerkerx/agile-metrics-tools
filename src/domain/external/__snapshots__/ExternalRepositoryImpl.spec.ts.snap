// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalRepositoryImpl should get inline values properly 1`] = `
Array [
  Object {
    "createdAt": 2018-12-04T00:00:00.000Z,
    "header1": 2,
    "header2": 3,
    "header3": "someValue",
    "teamName": "someTeamName",
  },
]
`;
